### PR TITLE
Land preview2-prototying's wasi-preview1-component-adapter into wasmtime

### PR DESCRIPTION
**Pull request by pchickey**
First task in #6370 

This introduces preview2-prototyping's `wasi-preview1-component-adapter` into wasmtime's `crates/`.  The git history has been preserved.

There is a new script `./ci/build-wasi-preview1-component-adapter.sh` to build and verify the adapter, and a a new CI check which runs this script and uploads the release builds as artifacts.

## How this PR was made

I made this branch by running
```
$ python3 ~/Downloads/git-filter-repo.py --path crates --path src --path build.rs --force
```
on https://github.com/bytecodealliance/preview2-prototyping/commit/5be99d645c177198b9f5e4d76561669796f7b379

And then merging wasmtime's `main` in with `--allow-unrelated-histories`.

### CODE REVIEW

**Comment by pchickey** [comment]
This PR is too big contains too many commits for `prtest:full` to work properly, so I temporarily removed the if test-full predicate in https://github.com/bytecodealliance/wasmtime/pull/6374/commits/bb5d051584658f24b5a3caa403726b619fb1257a and restored it in https://github.com/bytecodealliance/wasmtime/pull/6374/commits/968a61dbc402fbfc429353f9b4d30b479d1fa556

File: crates/wasi-preview1-component-adapter/Cargo.toml
Comment (alexcrichton) [inline crates/wasi-preview1-component-adapter/Cargo.toml:46]: Could this get removed so this is integrated into the main wasmtime workspace? I think the above fiddly bits about profiles and such can go into the main workspace as well and apply to just this one crate.

File: crates/wasi-preview1-component-adapter/.gitignore
Comment (alexcrichton) [inline crates/wasi-preview1-component-adapter/.gitignore:1]: Could this file get removed?

File: crates/wasi-preview1-component-adapter/wit/deps.lock
Comment (alexcrichton) [inline crates/wasi-preview1-component-adapter/wit/deps.lock:1]: Could this be left out until CI is added to verify it's right?

**Review by alexcrichton — APPROVED** [review]
Looks good to me, but I'd ideally like to get the workspace stuff sorted out before landing if that's ok. If this is holding up too much else though and that's not easy to get working I can work on it later and you can go ahead and land this.

File: crates/wasi-preview1-component-adapter/wit/deps.lock
Comment (pchickey) [reply crates/wasi-preview1-component-adapter/wit/deps.lock:1]: Yes, this should be left out until we actually setup wit-deps to point at the standards repos and check things properly in CI.

File: crates/wasi-preview1-component-adapter/.gitignore
Comment (pchickey) [reply crates/wasi-preview1-component-adapter/.gitignore:1]: Yes if we fix the workspace.

File: crates/wasi-preview1-component-adapter/Cargo.toml
Comment (pchickey) [reply crates/wasi-preview1-component-adapter/Cargo.toml:46]: I would love to put it all in one workspace, yeah. I wasn't able to figure out how to apply these profiles to just one crate, but if we can do it then

Comment (alexcrichton) [reply crates/wasi-preview1-component-adapter/Cargo.toml:46]: I think [profile overrides](https://doc.rust-lang.org/nightly/cargo/reference/profiles.html#overrides) show enable that?

Comment (pchickey) [reply crates/wasi-preview1-component-adapter/Cargo.toml:46]: That didn't work because
> Overrides cannot specify the `panic` ... settings
but I just realized, since that we have to avoid the panic machinery for vtable reasons, maybe the adapter will also build correctly without that in the profile...

Comment (alexcrichton) [reply crates/wasi-preview1-component-adapter/Cargo.toml:46]: oh that should also be ok because wasm doesn't actually support panic=unwind vs panic=abort, it's just always panic=abort anyway. So that setting can be dropped without affecting the output.

Comment (pchickey) [reply crates/wasi-preview1-component-adapter/Cargo.toml]: Thanks!

**Comment by pchickey** [comment]
Feedback is taken care of, and collapsing it into one workspace brought a bunch of `cargo vet` stuff into scope, so that is now taken care of as well.

**Comment by alexcrichton** [comment]
This [failed on CI](https://github.com/bytecodealliance/wasmtime/actions/runs/4984252582/jobs/8922338562), but additionally I think we'll want to temporarily set the merge strategy for the merge queue to creating a merge commit for merging this instead of the current squash-and-merge strategy

File: ci/run-tests.sh
Comment (alexcrichton) [inline ci/run-tests.sh:10]: Instead of this could the manifest of the adapter say `test = false` for the `[lib]` section?

**Comment by pchickey** [comment]
oof, good catch. After all that work to preserve the git history I almost goofed the squash commit.

**Pull request merged**
